### ISSUE TITLE
Get object templates from MISP instance in the format of misp-objects

### ISSUE THREAD

**Issue by chrisinmtown** [MISP/MISP#6591]
Extend MISP server with REST API get template by name, yielding a misp-object format.  This will allow the PyMISP client to query the MISP server for a custom object definition (e.g., "my-internal-bad-event") instead of requiring & reading a definition.json file locally in its site-packages area.

**Describe the solution you'd like**

Objects controller has a method to get an object template (maybe templates plural?) by name.

**Describe alternatives you've considered**

Today client code either requires a local copy of the object template definition in site packages, or must encode all the attribute types etc. Allowing fetch of server-side object template would eliminate both those awkward solutions.

**Additional context**

Please see PyMISP issue 659: https://github.com/MISP/PyMISP/issues/659

**Comment by chrisinmtown** [comment]
if one of you PHP wizards would sketch out the work briefly, I can try to work up a first draft.  Probably just a new function in  ObjectTemplatesController.php right?

**Comment by mokaddem** [comment]
You can search for template in the `objectTemplates` index by supplying the name in the `searchall` parameter.
You can search for `name`, `uuid`, `meta-category` and `description`
```
/objectTemplates/index/searchall:ail-leak
```
If you want to have the json version, just append `.json` at the end of the URL.
@chrisinmtown does that work around help?

**Comment by chrisinmtown** [comment]
Thanks @mokaddem. I opened this issue bcos @Rafiot posted a reply over in PyMISP issue 659 about getting an object template in misp-object format. Does that MISP feature exist already? Maybe Rafiot didn't know about it, and no work needs to be done?

**Comment by Rafiot** [comment]
Well, yes, I know about this one. But it is not in the format of the `misp-object` repository, so it is not something I can use directly as a template in PyMISP.

It is possible to create a thing that turns the template from MISP back into the original format, but why not just allowing to get a template in the proper format directly from MISP?

**Comment by chrisinmtown** [comment]
Just for the record, for beginners like myself, hitting URL https://misp/objectTemplates/index/searchall:blog.json in a browser yields the following object for the "blog" object template which is one of the two members in the list result, the other was "microblog":
```
 {
        "ObjectTemplate": {
            "id": "200",
            "user_id": "14",
            "org_id": "2",
            "uuid": "1f165fc0-b158-498f-8bc8-6dc3d2822bb1",
            "name": "blog",
            "meta-category": "misc",
            "description": "Blog post like Medium or WordPress.",
            "version": "14",
            "requirements": {
                "requiredOneOf": [
                    "post",
                    "archive",
                    "url",
                    "link"
                ]
            },
            "fixed": true,
            "active": true
        },
        "Organisation": {
            "id": "2",
            "name": "CSO",
            "uuid": "5a6a4161-b2e0-46b6-b12e-1bca0a6b037c"
        }
    }
```
But the misp-object format for this item from file /var/www/MISP/app/files/misp-objects/objects/blog/definition.json is below:
```
{
  "attributes": {
    "archive": {
      "description": "Archive of the original document (Internet Archive, Archive.is, etc).",
      "misp-attribute": "link",
      "multiple": true,
      "ui-priority": 1
    },
    "creation-date": {
      "description": "Initial creation of the blog post.",
      "misp-attribute": "datetime",
      "ui-priority": 0
    },
    "embedded-link": {
      "description": "Site linked by the blog post.",
      "misp-attribute": "url",
      "multiple": true,
      "ui-priority": 0
    },
    "embedded-safe-link": {
      "description": "Safe site linked by the blog post.",
      "misp-attribute": "link",
      "multiple": true,
      "ui-priority": 0
    },
    "link": {
      "description": "Original link into the blog post (Supposed harmless).",
      "misp-attribute": "link",
      "ui-priority": 1
    },
    "modification-date": {
      "description": "Last update of the blog post.",
      "misp-attribute": "datetime",
      "ui-priority": 0
    },
    "post": {
      "description": "Raw post.",
      "misp-attribute": "text",
      "ui-priority": 1
    },
    "removal-date": {
      "description": "When the blog post was removed.",
      "misp-attribute": "datetime",
      "ui-priority": 0
    },
    "title": {
      "description": "Title of blog post.",
      "misp-attribute": "text",
      "ui-priority": 1
    },
    "type": {
      "description": "Type of blog post.",
      "disable_correlation": true,
      "misp-attribute": "text",
      "sane_default": [
        "Medium",
        "WordPress",
        "Blogger",
        "Tumbler",
        "LiveJournal",
        "Forum",
        "Other"
      ],
      "ui-priority": 1
    },
    "url": {
      "description": "Original URL location of the blog post (potentially malicious).",
      "misp-attribute": "url",
      "ui-priority": 1
    },
    "username": {
      "description": "Username who posted the blog post.",
      "misp-attribute": "text",
      "ui-priority": 0
    },
    "username-quoted": {
      "description": "Username who are quoted into the blog post.",
      "misp-attribute": "text",
      "multiple": true,
      "ui-priority": 0
    },
    "verified-username": {
      "description": "Is the username account verified by the operator of the blog platform.",
      "disable_correlation": true,
      "misp-attribute": "text",
      "ui-priority": 0,
      "values_list": [
        "Verified",
        "Unverified",
        "Unknown"
      ]
    }
  },
  "description": "Blog post like Medium or WordPress.",
  "meta-category": "misc",
  "name": "blog",
  "requiredOneOf": [
    "post",
    "archive",
    "url",
    "link"
  ],
  "uuid": "1f165fc0-b158-498f-8bc8-6dc3d2822bb1",
  "version": 14
}
```

**Comment by chrisinmtown** [comment]
Hello CIRCL team, what is the right method for the server to retrieve the misp-object format for a named object template?  Just open the file at the expected path and read it, or should a cache or the database be used instead?

**Comment by mokaddem** [comment]
@chrisinmtown I'm not sure to understand the question.
Whenever your do a full update for your template, MISP will do the following:
1. Pull from the object-template repository
2. Read the file on the disk and load it into the database

I hope that helps.

**Comment by Rafiot** [comment]
@chrisinmtown: a new API endpoint will be created specially to get the **latest** version of an object template definition from a MISP instance, by UUID.

**Comment by chrisinmtown** [comment]
Thanks @Rafiot. For PyMISP purposes, if I would like to create a new object based on a server-side (not local) template, I think the client will only have the object template name, not its UUID.  It would be very convenient for PyMISP implementation if the server-side method allowed fetch of object template by exact match on name in addition to UUID.  Please consider that.

I was asking for the recommended approach so I could try to contribute the new method, or at least a first draft.

**Comment by Rafiot** [comment]
Getting a template by name will have to return a list, as it is possible (even if discouraged) to have templates with the same name, but a different UUID.

**Comment by chrisinmtown** [comment]
Thanks for PR https://github.com/MISP/MISP/pull/6670 would you please move ahead with it?

**Comment by chrisinmtown** [comment]
PR #6670 has been merged to branch develop, you all want to keep this open until that branch is merged to 2.4?
